Add blog post about sbt-dependency-submission (#1398)
scala/scala-lang: blog/_posts/2022-07-18-secure-your-dependencies-on-github.md

Being able to identify new vulnerabilities and to respond to them quickly reduces the risk of exposure to its minimum.
But the task can be time consuming, unless we automate it.

GitHub Dependabot is an automation tool that helps us secure our projects.
Its initial approach is to scan static configuration files such as POM files or package-lock.json files to extract the dependencies of a repository and cross reference them with a database of vulnerabilities.
But some build definitions, such as build.sbt files, are programs that cannot be statically analyzed.
A few weeks ago, GitHub released the beta version of the Dependency submission API, a REST API that we can use in CI workflows to submit complete snapshots of dependencies when they are resolved at build-time.

As a large part of the Scala community uses GitHub to host their projects, we developed sbt-dependency-submission, a GitHub action that can extract the graphs of dependencies of sbt projects, for Dependabot to analyze them and report any vulnerability.
With this tool, we hope to contribute to building a more secure Scala open source ecosystem.

## Setting up the workflow

> Before installing the dependency submission action, you need to ensure that the `Dependency Graph` view is enabled in the `Settings > Code security and analysis` tab of your GitHub repository.

Reviewer: ```suggestion
Before installing the dependency submission action, you need to ensure that the `Dependency graph` view is enabled in the `Settings > Code security and analysis` tab of your GitHub repository.
```

You should also consider enabling `Dependabot alerts`.

![Dependency Graph](/resources/img/blog/github/dependency-graph.jpg)

To install the action, add a new workflow in the `.github/workflows` folder, with the following definition:

```yaml
# .github/workflows/dependency-graph.yml
name: Update Dependency Graph

on:
  push:
    branches:
      - main # default branch of the project

> # The API requires write permission on the repository to submit dependencies

Reviewer 2: This is by default though isn't it? For example in the mill action I don't have any permissions set and it still works fine.

permissions:
  contents: write

jobs:
  update-graph:
    name: Update Dependency Graph
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: scalacenter/sbt-dependency-submission@v2
    with:
      ## Optional: Define the working directory of your build.
      ## It should contain the build.sbt file.
 	    working-directory: ‘./’
```

This default definition should work in many sbt builds out-of-the-box.

> The action will check that the working directory contains an sbt project, it will install an sbt plugin, and run a command, added by that plugin, that extracts all resolved dependencies from all Scala projects on all Scala versions.

Reviewer 3: nit: I feel like there's too much mechanical details in this sentence that can be omitted.

```suggestion
Given a valid sbt build, the action installs an sbt plugin that will extract all transitive library dependencies from all subprojects on all Scala versions.
```

If you need to exclude a project on a Scala version, because sbt cannot resolve its dependencies, you can use the `modules-ignore` input.
You can learn more about this in the [documentation page](https://github.com/marketplace/actions/sbt-dependency-submission) of the action.

## Scala Steward and GitHub Dependabot

[Scala Steward](https://github.com/scala-steward-org/scala-steward) is a tool that helps you keep the dependencies of your project up-to-date by opening pull requests on GitHub (and other hosting services).
It is used in more than a thousand open-source repositories and many proprietary ones and contributes to a large extent to the security of the Scala ecosystem.

Dependabot and Scala Steward can be used as complementary tools.
Scala Steward, as a preventive tool, can help you keep your dependencies up-to-date, which reduces the risk of security vulnerabilities.

> Dependabot, as a monitoring tool, can notify you when a vulnerability is found, so that you can act quickly.

Author: Well, that does not yet work with sbt or Mill

> Dependabot, as a monitoring tool, can notify you when a vulnerability is found, so that you can act quickly.

Reviewer 2: Sure, but it does for your github actions, which is important here. For example if you bump `sbt-dependency-submission` to `v3` a user won't know that unless they have dependabot set up to update their actions.

## Other supported build tools

You can use Dependabot with other Scala-compatible build tools.

### Maven

Dependabot supports Maven through static analysis of POM files.
You can enable the Dependency graph view and Dependabot in the `Settings > Code security and analysis` page.

### Mill
The [mill-dependency-submission][mill-dependency-submisison] action, developed by Chris Kipp, can extract the dependencies of a Mill build.

To install it:

>  - Enable the `Dependency Graph` view, and optionally the `Dependabot alerts`, in the `Settings > Code security and analysis` page of your repository.

Reviewer: ```suggestion
 - Enable the `Dependency graph` view, and optionally the `Dependabot alerts`, in the `Settings > Code security and analysis` page of your repository.
```

 - Add this workflow in you repository:

```yaml
# .github/workflows/dependency-graph.yml
name: Update Dependency Graph

on:
  push:
    branches:
      - main # default branch of the project

# The API requires write permission on the repository to submit dependencies
permissions:
  contents: write

jobs:
  update-graph:
    name: Update Dependency Graph
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: coursier/cache-action@v6
      - uses: actions/setup-java@v3
        with:
          distribution: 'temurin'
          java-version: '17'
      - uses: ckipp01/mill-dependency-submission@v1
```
### Gradle

Gradle users can use the [gradle-dependency-submission][gradle-dependency-submisison] action.

## Summary

> In Github, it is now possible to use pre-made actions, like [sbt-dependency-submisison][sbt-dependency-submisison] or [mill-dependency-submisison][mill-dependency-submisison], to extract full graphs of dependencies at build time.

Reviewer 3: There's a typo in "submission" as submisison both in the text and the Markdown link name.
